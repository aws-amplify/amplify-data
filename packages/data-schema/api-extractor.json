{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
