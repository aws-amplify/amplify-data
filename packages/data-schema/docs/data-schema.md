<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
