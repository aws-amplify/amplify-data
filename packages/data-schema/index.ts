import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
