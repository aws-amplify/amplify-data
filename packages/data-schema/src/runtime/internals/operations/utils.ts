// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
// import { GraphQLFormattedError } from '@aws-amplify/data-schema-types';

/**
 * Handle errors for list return types (list and index query operations)
 */
export function handleListGraphQlError(error: any) {
  if (error?.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: [],
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 * Handle errors for singular return types (create, get, update, delete operations)
 */
export function handleSingularGraphQlError(error: any) {
  if (error.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: null,
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 *
 * @param resolver
 * @returns
 */
export function selfAwareAsync<T>(
  resolver: (promise: Promise<T>) => Promise<T>,
): Promise<T> {
  let resolve: (value: T) => void;
  let reject: (reason?: any) => void;

  const resultPromise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  resolver(resultPromise)
    .then((result) => {
      resolve(result);
    })
    .catch((error) => {
      reject(error);
    });

  return resultPromise;
}
