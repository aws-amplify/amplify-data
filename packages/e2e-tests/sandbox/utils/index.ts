import { deploySandbox, deploySandboxWatch } from './deploy';
import { teardownSandbox } from './teardown';
import { updateSchema } from './schemaUtils';

/**
 * Sandbox operations can sometimes be time consuming, so we increase the Jest
 * setup / teardown hook timeouts to 10 minutes.
 */
export const sandboxTimeout: number = 600000;

export const outputsFileName = 'amplify_outputs.json';

export { deploySandbox, deploySandboxWatch };
export { teardownSandbox };
export { updateSchema };
