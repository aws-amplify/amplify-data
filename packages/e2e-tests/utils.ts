import { Amplify } from 'aws-amplify';
import { generateClient } from 'aws-amplify/data';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';

export const statuses = {
  success: {
    icon: '✅',
    label: 'Success',
  },
  fail: {
    icon: '❌',
    label: 'Fail',
  },
} as const;

const _testCase = (success: boolean): any =>
  success ? statuses.success : statuses.fail;

// TODO: use imported type from `aws-amplify/data` once it's fixed
type Client = ReturnType<typeof configureAmplifyAndGenerateClient>;

export type TestCase = {
  label: string;
  action: (client: Client) => Promise<boolean>;
};

/**
 * Given an array of test cases, runs each test case
 * and asserts that the result is successful
 * @param testCases
 * @param client
 */
export async function runTestCases(testCases: TestCase[]) {
  const client = (global as any).client as Client;

  for (const testCase of testCases) {
    test(`${testCase.label}`, async () => {
      let result: any;

      try {
        const response = await testCase.action(client);
        result = _testCase(response);
      } catch (error) {
        result = _testCase(false);
      }

      console.log(`${result.icon} ${testCase.label}`);
      expect(result.label).toBe(statuses.success.label);
    });
  }
}

/**
 * Configures Amplify and returns API client
 * @returns API client
 */
export function configureAmplifyAndGenerateClient() {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs);

  const client = generateClient<Schema>();

  return client;
}
