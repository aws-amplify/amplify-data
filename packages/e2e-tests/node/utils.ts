import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
// TODO: the dep exists, investigate why it's not being resolved:
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  disableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param disableDebugLogging - disables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  disableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export async function pause(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type EstablishWebsocketParams = {
  disableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const establishWebsocket = ({
  disableConnectionStateLogging = false,
}: EstablishWebsocketParams): any => {
  (global as any).WebSocket = WebSocket;

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableConnectionStateLogging) {
    // https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/#subscription-connection-status-updates
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

export const handleErrorsAndData = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};
