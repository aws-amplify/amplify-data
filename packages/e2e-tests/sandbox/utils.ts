import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  enableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param enableDebugLogging - enables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  enableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  if (enableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

type AddWebsocketPolyfillParams = {
  enableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const addWebsocketPolyfill = ({
  enableConnectionStateLogging = false,
}: AddWebsocketPolyfillParams) => {
  (global as any).WebSocket = WebSocket;

  if (enableConnectionStateLogging) {
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export const expectDataReturnWithoutErrors = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};

/**
 * Function that listens for "Subscription ack" event, and once it receives it,
 * resolves.
 * Jest will timeout if the test does not complete within the default timeout of
 * 5 seconds, so no timeout handling is required.
 */
export const waitForSubscriptionAck = () =>
  new Promise<void>((resolve) => {
    const cancel = Hub.listen('api', async (data: any) => {
      const { payload } = data;
      // TODO: update once we add export:
      if (payload.event === 'Subscription ack') {
        console.log('Subscription ack received:', payload.data);
        // Stop listening for msgs once we receive the ack:
        cancel();
        // Resolve so test execution can continue:
        resolve();
      }
    });
  });
