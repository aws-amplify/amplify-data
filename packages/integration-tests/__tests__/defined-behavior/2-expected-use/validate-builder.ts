import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { buildAmplifyConfig, mockedGenerateClient, expectVariables } from '../../utils';
import { GraphQLError } from 'graphql';

describe('Validate Builder', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Validate Builder', () => {
    const schema = a.schema({
      Post: a.model({
        title: a.string().validate(v =>
          v.minLength(5, 'Title must be at least 5 characters')
           .maxLength(200, 'Title must be at most 200 characters')
        ),
        slug: a.string().validate(v =>
          v.matches('^[a-z0-9-]+$', 'Slug must contain only lowercase letters, numbers, and hyphens')
        ),
        summary: a.string().validate(v =>
          v.maxLength(500, 'Summary must be at most 500 characters')
        ),
        content: a.string().validate(v =>
          v.minLength(20, 'Content must be at least 20 characters')
        ),
      }).authorization((allow) => allow.publicApiKey()),
    
      Coupon: a.model({
        code: a.string().validate(v =>
          v.startsWith('CPN-', 'Code must start with CPN-')
           .endsWith('-2025', 'Code must end with -2025')
        ),
      }).authorization((allow) => allow.publicApiKey()),
    
      Product: a.model({
        price: a.float().validate(v => 
          v.positive('Price must be positive')
           .lt(10000, 'Price must be less than 10000')
        ),
        rating: a.float().validate(v =>
          v.gte(0, 'Rating must be at least 0')
           .lte(5, 'Rating must be at most 5')
        ),
        stock: a.integer().validate(v =>
          v.gte(0, 'Stock must be at least 0')
           .lte(1000, 'Stock must be at most 1000')
        ),
        profit: a.float().validate(v =>
          v.gt(0, 'Profit must be greater than 0')
        ),
        loss: a.float().validate(v =>
          v.negative('Loss must be negative')
        ),
      }).authorization((allow) => allow.publicApiKey()),
    });

    //===========================================
    // SCHEMA TRANSFORMATION TESTS
    //===========================================
    it('generates expected graphql', () => {
      const graphql = schema.transform().schema;
      expect(graphql).toMatchSnapshot();  
    });

    type Schema = ClientSchema<typeof schema>;
    
    const createMockResponse = (modelName: string, data: any, errors?: GraphQLError[]) => ({
      data: { [`create${modelName}`]: data },
      errors
    });
    
    const updateMockResponse = (modelName: string, data: any, errors?: GraphQLError[]) => ({
      data: { [`update${modelName}`]: data },
      errors
    });
    
    // Mock responses organized by model
    const mockResponses = [
      // Post responses
      createMockResponse('Post', { 
        id: 'post-id-1', 
        title: 'Valid title', 
        slug: 'valid-slug-123', 
        summary: 'A valid summary', 
        content: 'This is valid content that is longer than 20 characters.' 
      }),
      updateMockResponse('Post', { 
        id: 'post-id-1', 
        title: 'Updated title', 
        slug: 'updated-slug-123', 
        summary: 'An updated summary', 
        content: 'This is updated content that is longer than 20 characters.' 
      }),
      createMockResponse('Post', { 
        id: 'post-id-short-title', 
        title: 'Hi', // Too short, violates minLength(5)
        slug: 'valid-slug',
        summary: 'Summary',
        content: 'Long enough content here'
      }, [new GraphQLError('Validation error: Title must be at least 5 characters')]),
      updateMockResponse('Post', {
        id: 'post-id-invalid-slug', 
        title: 'Valid title',
        slug: 'INVALID_SLUG', // Contains uppercase and underscore
        summary: 'Summary',
        content: 'Long enough content here'
      }, [new GraphQLError('Validation error: Slug must contain only lowercase letters, numbers, and hyphens')]),
      createMockResponse('Post', { 
        id: 'post-id-long-summary', 
        title: 'Valid title',
        slug: 'valid-slug',
        summary: 'A'.repeat(501), // Too long
        content: 'Long enough content here'
      }, [new GraphQLError('Validation error: Summary must be at most 500 characters')]),
      updateMockResponse('Post', {
        id: 'post-id-short-content', 
        title: 'Valid title',
        slug: 'valid-slug',
        summary: 'Summary',
        content: 'Too short' // Too short
      }, [new GraphQLError('Validation error: Content must be at least 20 characters')]),
      
      // Coupon responses
      createMockResponse('Coupon', {
        id: 'coupon-id-1',
        code: 'CPN-SUMMER-2025'
      }),
      updateMockResponse('Coupon', {
        id: 'coupon-id-1',
        code: 'CPN-WINTER-2025'
      }),
      createMockResponse('Coupon', {
        id: 'coupon-id-wrong-prefix',
        code: 'COUPON-SUMMER-2025' // Wrong prefix
      }, [new GraphQLError('Validation error: Code must start with CPN-')]),
      updateMockResponse('Coupon', {
        id: 'coupon-id-wrong-suffix',
        code: 'CPN-SUMMER-2024' // Wrong suffix
      }, [new GraphQLError('Validation error: Code must end with -2025')]),
      
      // Product responses
      createMockResponse('Product', {
        id: 'product-id-1',
        price: 99.99,
        rating: 4.5,
        stock: 100,
        profit: 25.50,
        loss: -10.25
      }),
      updateMockResponse('Product', {
        id: 'product-id-1',
        price: 129.99,
        rating: 4.8,
        stock: 75,
        profit: 35.50,
        loss: -5.25
      }),
      updateMockResponse('Product', {
        id: 'product-id-negative-price',
        price: -10.00, // Negative
        rating: 4.5,
        stock: 100,
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Price must be positive')]),
      createMockResponse('Product', {
        id: 'product-id-high-price',
        price: 15000.00, // Too high
        rating: 4.5,
        stock: 100,
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Price must be less than 10000')]),
      updateMockResponse('Product', {
        id: 'product-id-negative-rating',
        price: 99.99,
        rating: -1.0, // Below minimum
        stock: 100,
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Rating must be at least 0')]),
      createMockResponse('Product', {
        id: 'product-id-high-rating',
        price: 99.99,
        rating: 6.0, // Above maximum
        stock: 100,
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Rating must be at most 5')]),
      updateMockResponse('Product', {
        id: 'product-id-negative-stock',
        price: 99.99,
        rating: 4.5,
        stock: -10, // Negative
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Stock must be at least 0')]),
      createMockResponse('Product', {
        id: 'product-id-high-stock',
        price: 99.99,
        rating: 4.5,
        stock: 1500, // Too high
        profit: 25.50,
        loss: -10.25
      }, [new GraphQLError('Validation error: Stock must be at most 1000')]),
      createMockResponse('Product', {
        id: 'product-id-zero-profit',
        price: 99.99,
        rating: 4.5,
        stock: 100,
        profit: 0.0, // Not positive
        loss: -10.25
      }, [new GraphQLError('Validation error: Profit must be greater than 0')]),
      updateMockResponse('Product', {
        id: 'product-id-positive-loss',
        price: 99.99,
        rating: 4.5,
        stock: 100,
        profit: 25.50,
        loss: 10.25 // Positive
      }, [new GraphQLError('Validation error: Loss must be negative')]),
    ];
    
    const { spy, generateClient } = mockedGenerateClient(mockResponses);
    let client: ReturnType<typeof generateClient<Schema>>;
    
    beforeAll(async () => {
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      client = generateClient<Schema>();
    });
    
    const testValidCreate = async (modelName: string, data: any) => {
      await (client.models as any)[modelName].create(data);
      expectVariables(spy, { input: data });
    };
    
    const testValidUpdate = async (modelName: string, data: any) => {
      await (client.models as any)[modelName].update(data);
      expectVariables(spy, { input: data });
    };
    
    const testInvalidOperation = async (operation: 'create' | 'update', modelName: string, data: any, errorMessage: string) => {
      await (client.models as any)[modelName][operation](data)
        .catch((error: any) => {
          expect(error.message).toContain(errorMessage);
        });
    };
    
    //===========================================
    // POST MODEL TESTS
    //===========================================
    describe('Post Model Validation', () => {
      describe('Create and Update Operations', () => {
        it('should successfully create a Post with valid fields', async () => {
          await testValidCreate('Post', {
            title: 'Valid title',
            slug: 'valid-slug-123',
            summary: 'A valid summary',
            content: 'This is valid content that is longer than 20 characters.'
          });
        });
        
        it('should successfully update a Post with valid fields', async () => {
          await testValidUpdate('Post', {
            id: 'post-id-1',
            title: 'Updated title',
            slug: 'updated-slug-123',
            summary: 'An updated summary',
            content: 'This is updated content that is longer than 20 characters.'
          });
        });
        
        it('should reject a Post with title too short', async () => {
          await testInvalidOperation('create', 'Post', {
            title: 'Hi', // Too short
            slug: 'valid-slug',
            summary: 'Summary',
            content: 'Long enough content here'
          }, 'Title must be at least 5 characters');
        });
        
        it('should reject a Post update with invalid slug pattern', async () => {
          await testInvalidOperation('update', 'Post', {
            id: 'post-id-invalid-slug',
            title: 'Valid title',
            slug: 'INVALID_SLUG', // Invalid pattern
            summary: 'Summary',
            content: 'Long enough content here'
          }, 'Slug must contain only lowercase letters, numbers, and hyphens');
        });
        
        it('should reject a Post with summary too long', async () => {
          await testInvalidOperation('create', 'Post', {
            title: 'Valid title',
            slug: 'valid-slug',
            summary: 'A'.repeat(501), // Too long
            content: 'Long enough content here'
          }, 'Summary must be at most 500 characters');
        });
        
        it('should reject a Post update with content too short', async () => {
          await testInvalidOperation('update', 'Post', {
            id: 'post-id-short-content',
            title: 'Valid title',
            slug: 'valid-slug',
            summary: 'Summary',
            content: 'Too short' // Too short
          }, 'Content must be at least 20 characters');
        });
      });
    });
    
    //===========================================
    // COUPON MODEL TESTS
    //===========================================
    describe('Coupon Model Validation', () => {
      describe('Create and Update Operations', () => {
        it('should successfully create a Coupon with valid code', async () => {
          await testValidCreate('Coupon', {
            code: 'CPN-SUMMER-2025'
          });
        });
        
        it('should successfully update a Coupon with valid code', async () => {
          await testValidUpdate('Coupon', {
            id: 'coupon-id-1',
            code: 'CPN-WINTER-2025'
          });
        });
        
        it('should reject a Coupon with invalid prefix', async () => {
          await testInvalidOperation('create', 'Coupon', {
            code: 'COUPON-SUMMER-2025' // Wrong prefix
          }, 'Code must start with CPN-');
        });
        
        it('should reject a Coupon update with invalid suffix', async () => {
          await testInvalidOperation('update', 'Coupon', {
            id: 'coupon-id-wrong-suffix',
            code: 'CPN-SUMMER-2024' // Wrong suffix
          }, 'Code must end with -2025');
        });
      });
    });
    
    //===========================================
    // PRODUCT MODEL TESTS
    //===========================================
    describe('Product Model Validation', () => {
      describe('Create and Update Operations', () => {
        it('should successfully create a Product with valid fields', async () => {
          await testValidCreate('Product', {
            price: 99.99,
            rating: 4.5,
            stock: 100,
            profit: 25.50,
            loss: -10.25
          });
        });
        
        it('should successfully update a Product with valid fields', async () => {
          await testValidUpdate('Product', {
            id: 'product-id-1',
            price: 129.99,
            rating: 4.8,
            stock: 75,
            profit: 35.50,
            loss: -5.25
          });
        });
        
        it('should reject a Product update with negative price', async () => {
          await testInvalidOperation('update', 'Product', {
            id: 'product-id-negative-price',
            price: -10.00, // Negative
            rating: 4.5,
            stock: 100,
            profit: 25.50,
            loss: -10.25
          }, 'Price must be positive');
        });
        
        it('should reject a Product with price too high', async () => {
          await testInvalidOperation('create', 'Product', {
            price: 15000.00, // Too high
            rating: 4.5,
            stock: 100,
            profit: 25.50,
            loss: -10.25
          }, 'Price must be less than 10000');
        });
        
        it('should reject a Product update with negative rating', async () => {
          await testInvalidOperation('update', 'Product', {
            id: 'product-id-negative-rating',
            price: 99.99,
            rating: -1.0, // Below minimum
            stock: 100,
            profit: 25.50,
            loss: -10.25
          }, 'Rating must be at least 0');
        });
        
        it('should reject a Product with rating too high', async () => {
          await testInvalidOperation('create', 'Product', {
            price: 99.99,
            rating: 6.0, // Above maximum
            stock: 100,
            profit: 25.50,
            loss: -10.25
          }, 'Rating must be at most 5');
        });
        
        it('should reject a Product update with negative stock', async () => {
          await testInvalidOperation('update', 'Product', {
            id: 'product-id-negative-stock',
            price: 99.99,
            rating: 4.5,
            stock: -10, // Negative
            profit: 25.50,
            loss: -10.25
          }, 'Stock must be at least 0');
        });
        
        it('should reject a Product with stock too high', async () => {
          await testInvalidOperation('create', 'Product', {
            price: 99.99,
            rating: 4.5,
            stock: 1500, // Too high
            profit: 25.50,
            loss: -10.25
          }, 'Stock must be at most 1000');
        });
        
        it('should reject a Product with zero profit', async () => {
          await testInvalidOperation('create', 'Product', {
            price: 99.99,
            rating: 4.5,
            stock: 100,
            profit: 0.0, // Not positive
            loss: -10.25
          }, 'Profit must be greater than 0');
        });
        
        it('should reject a Product update with positive loss', async () => {
          await testInvalidOperation('update', 'Product', {
            id: 'product-id-positive-loss',
            price: 99.99,
            rating: 4.5,
            stock: 100,
            profit: 25.50,
            loss: 10.25 // Positive
          }, 'Loss must be negative');
        });
      });
    });
  });
});