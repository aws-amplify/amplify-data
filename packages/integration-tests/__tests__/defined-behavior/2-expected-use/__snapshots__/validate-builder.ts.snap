// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Builder Validate Builder generates expected graphql 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String @validate(type: minLength, value: "5", errorMessage: "Title must be at least 5 characters") @validate(type: maxLength, value: "200", errorMessage: "Title must be at most 200 characters")
  slug: String @validate(type: matches, value: "^[a-z0-9-]+$", errorMessage: "Slug must contain only lowercase letters, numbers, and hyphens")
  summary: String @validate(type: maxLength, value: "500", errorMessage: "Summary must be at most 500 characters")
  content: String @validate(type: minLength, value: "20", errorMessage: "Content must be at least 20 characters")
}

type Coupon @model @auth(rules: [{allow: public, provider: apiKey}])
{
  code: String @validate(type: startsWith, value: "CPN-", errorMessage: "Code must start with CPN-") @validate(type: endsWith, value: "-2025", errorMessage: "Code must end with -2025")
}

type Product @model @auth(rules: [{allow: public, provider: apiKey}])
{
  price: Float @validate(type: gt, value: "0", errorMessage: "Price must be positive") @validate(type: lt, value: "10000", errorMessage: "Price must be less than 10000")
  rating: Float @validate(type: gte, value: "0", errorMessage: "Rating must be at least 0") @validate(type: lte, value: "5", errorMessage: "Rating must be at most 5")
  stock: Int @validate(type: gte, value: "0", errorMessage: "Stock must be at least 0") @validate(type: lte, value: "1000", errorMessage: "Stock must be at most 1000")
  profit: Float @validate(type: gt, value: "0", errorMessage: "Profit must be greater than 0")
  loss: Float @validate(type: lt, value: "0", errorMessage: "Loss must be negative")
}"
`;
